src/main/cpp/src/hive_hash.cu

```diff
@@ -363,53 +392,57 @@ class hive_device_row_hasher {
                                             cudf::size_type row_index) const noexcept
     {
-      cudf::column_device_view curr_col = col.slice(row_index, 1);
-      // The default constructor of `col_stack_frame` is deleted, so it can not allocate an array
-      // of `col_stack_frame` directly.
-      // Instead leverage the byte array to create the col_stack_frame array.
-      alignas(col_stack_frame) char stack_wrapper[sizeof(col_stack_frame) * MAX_STACK_DEPTH];
-      auto col_stack = reinterpret_cast<col_stack_frame*>(stack_wrapper);
-      int stack_size = 0;
+      auto curr_col_idx = _nested_column_map[_col_idx];
+      auto curr_row_idx = row_index;
 
-      col_stack[stack_size++] = col_stack_frame(curr_col);
+      col_stack_frame col_stack[MAX_STACK_DEPTH];
+      int stack_size          = 0;
+      col_stack[stack_size++] = col_stack_frame(curr_col_idx, curr_row_idx);
 
       while (stack_size > 0) {
         col_stack_frame& top = col_stack[stack_size - 1];
-        curr_col             = top.get_column();
+        curr_col_idx         = top.get_col_idx();
+        curr_row_idx         = top.get_row_idx();
+        auto const& curr_col = _flattened_column_views[curr_col_idx];
         // Do not pop it until it is processed. The definition of `processed` is:
         // - For structs, it is when all child columns are processed.
         // - For lists, it is when all elements in the list are processed.
         if (curr_col.type().id() == cudf::type_id::STRUCT) {
           if (top.get_idx_to_process() == curr_col.num_child_columns()) {
             if (--stack_size > 0) { col_stack[stack_size - 1].update_cur_hash(top.get_hash()); }
           } else {
-            auto const structcv = cudf::detail::structs_column_device_view(curr_col);
             while (top.get_idx_to_process() < curr_col.num_child_columns()) {
-              auto idx             = top.get_and_inc_idx_to_process();
-              auto const child_col = structcv.get_sliced_child(idx);
+              auto idx              = top.get_and_inc_idx_to_process();
+              auto child_col_idx    = _first_child_index[curr_col_idx] + idx;
+              auto const& child_col = _flattened_column_views[child_col_idx];
               // If the child is of primitive type, accumulate child hash into struct hash
               if (child_col.type().id() != cudf::type_id::LIST &&
                   child_col.type().id() != cudf::type_id::STRUCT) {
                 auto child_hash =
                   cudf::type_dispatcher<cudf::experimental::dispatch_void_if_nested>(
-                    child_col.type(), this->hash_functor, child_col, 0);
+                    child_col.type(), this->hash_functor, child_col, curr_row_idx);
                 top.update_cur_hash(child_hash);
               } else {
-                col_stack[stack_size++] = col_stack_frame(child_col);
+                col_stack[stack_size++] = col_stack_frame(child_col_idx, curr_row_idx);
```

Comment:
IMHO, it is better to update the array element(col_stack_frame) inplacely instead of creating a new object.